### PR TITLE
Update boto3 and s3fs dependencies, drop Python 3.7 support

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
This PR updates boto3 and s3fs versions to their latest versions. The current versions conflict with various AWS APIs.

It also drops support for Python 3.7, which is
- Almost a year past EOL
- Not compatible with the latest boto3 version

Also updates daft.

## Note on benchmarks

Updating S3Fs introduced a regression in some of the benchmarks -- the mechanism for this is unknown at the moment, and it's not clear if this would translate to an actual performance regression in production. We're proceeding with caution.

### CODE REVIEW

File: deltacat/__init__.py [hunk 1]
```diff
@@ -44,7 +44,7 @@
 
 deltacat.logs.configure_deltacat_logger(logging.getLogger(__name__))
 
-__version__ = "1.1.7"
+__version__ = "1.1.8"
```
Comment (raghumdani) [inline deltacat/__init__.py:47]: Let's not bump yet. I have some other changes to go in today.

File: requirements.txt [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # setup.py install_requires
 # any changes here should also be reflected in setup.py "install_requires"
 aws-embedded-metrics == 3.2.0
-boto3 ~= 1.20
+boto3 ~= 1.34
 getdaft==0.2.23
```
Comment (raghumdani) [inline requirements.txt:5]: Do you want to upgrade daft to `0.2.25` too?

**Review by raghumdani — COMMENTED** [review]
Looks good mostly. Please address few comments.

File: deltacat/__init__.py (hunk 1 above)
Comment (DmitriGekhtman) [reply deltacat/__init__.py:47]: Sg, let's coordinate on the version bump.

File: requirements.txt (hunk 2 above)
Comment (DmitriGekhtman) [reply requirements.txt:5]: Will do.

File: deltacat/__init__.py (hunk 1 above)
Comment (DmitriGekhtman) [reply deltacat/__init__.py]: Reverted version bump.

File: requirements.txt (hunk 2 above)
Comment (DmitriGekhtman) [reply requirements.txt]: Upgraded daft.

**Comment by DmitriGekhtman** [comment]
Test failure after daft bump -- may or may not be related -- will look into it.

**Comment by DmitriGekhtman** [comment]
> [Py3.8 test failure](https://github.com/ray-project/deltacat/actions/runs/9289897697/job/25564867260?pr=309) after daft bump -- may or may not be related -- will look into it.

It succeeded on re-run. Looks like there might be a bit of test flakiness, specifically in `test_resources.py`.

**Comment by raghumdani** [comment]
There seems to be performance regression. Can we know if it's a false positive?

**Comment by DmitriGekhtman** [comment]
> There seems to be performance regression. Can we know if it's a false positive?

Looks like we're re-running the CI to find out.

**Comment by DmitriGekhtman** [comment]
Triggered 3 runs against this branch in
https://github.com/ray-project/deltacat/pull/310
https://github.com/ray-project/deltacat/pull/311
https://github.com/ray-project/deltacat/pull/312

Triggered 3 runs against master in
https://github.com/ray-project/deltacat/pull/313
https://github.com/ray-project/deltacat/pull/314
https://github.com/ray-project/deltacat/pull/315

We can compare results for the test of interest `test_compact_partition_rebase_then_incremental[9-rebase-then-incremental-single-hash-bucket_V2`

**Comment by DmitriGekhtman** [comment]
Already seeing some perf regression alerts from the runs against this branch.

**Comment by DmitriGekhtman** [comment]
Ok...results from round 3 seem to have overwritten the others, clearly I don't know how GitHub Actions works, that's ok. Anyways:

Against this branch:
Py39: 1.7277032199950095 iter/s, stdev 08763625144110297
Py310: 1.6111781171268102 iter/s, stdev 0.096839517459092

Against master
Py39: 3.0060425923086536, 0.06733375464572759
Py310: 3.2338998534936736, 0.08521136159332987

**Comment by DmitriGekhtman** [comment]
...downgrading daft does not appear to fix it

**Comment by DmitriGekhtman** [comment]
that leaves s3fs, let me see if downgrading that fixes it (sad)

**Comment by DmitriGekhtman** [comment]
> that leaves s3fs, let me see if downgrading that fixes it (sad)

looks like to downgrade s3fs, I would also have to downgrade boto3 to avoid a dependency conflict, which negates this PR entirely
let's back up to consider the options

**Comment by raghumdani** [comment]
Let's take the risk and observe real perf diff in prod.

**Review by raghumdani — APPROVED** [review]
LGTM

**Pull request merged**